[Index] Avoid forming relations to non-indexed decls

hamishknight opened the pull request:
Make sure to check whether we are able to index a decl before reporting it as a relation. Additionally for containers, walk up the stack to find the first container that we can index, avoiding treating local variables as containers when we're not indexing them.

rdar://126137541

hamishknight commented:
@swift-ci please test

hamishknight commented:
@swift-ci please test

hamishknight commented:
@swift-ci please test

bnbarham commented on lib/Index/Index.cpp:
```diff
@@ -216,26 +216,38 @@ class ContainerTracker {
 
-    if (!Entry.ActiveKey)
-      return;
+      SmallVector<const Decl *, 4> Containers;
```
Could have a bool for whether we actually allowed any decls or not and then just use `f(D)` in the two if's, but if you prefer this I'd move the `Containers` out of the loop.

hamishknight replied:
Sure, don't feel too strongly, switched to using a bool

hamishknight committed:
[Index] Avoid forming relations to non-indexed decls

If we shouldn't index the related decl, don't record
it as a relation.

hamishknight committed:
[Index] Avoid reporting containers for non-indexed decls

Check to see whether we can index the given decl
before reporting it as a container, walking up to
a parent if we need to. This also lets us simplify
the AnyPattern handling a bit.

rdar://126137541

hamishknight commented:
@swift-ci please test
